skada/_subspace.py

```diff
@@ -640,3 +642,322 @@ def TransferJointMatching(
+        """Compute the covariance matrix of X using torch."""
+        torch = self.torch
+        reg = self.reg
+
+        n_samples, d = X.shape
+        X = X - torch.mean(X, dim=0)
+        cov = X.T @ X / n_samples
+        cov = (1 - reg) * cov + reg * torch.trace(cov) * torch.eye(d)
+
+        return cov
+
+    def _D(self, W, X_source, X_target):
+        """Divergence objective function"""
+        torch = self.torch
+
+        Z_source = X_source @ W
+        Z_target = X_target @ W
+
+        sigma_1 = self._torch_cov(Z_source)
+        sigma_2 = self._torch_cov(Z_target)
+        sigma_11 = 2 * sigma_1
+        sigma_12 = sigma_1 + sigma_2
+        sigma_22 = 2 * sigma_2
+
+        L_11 = torch.linalg.cholesky(torch.linalg.inv(sigma_11))
+        L_12 = torch.linalg.cholesky(torch.linalg.inv(sigma_12))
+        L_22 = torch.linalg.cholesky(torch.linalg.inv(sigma_22))
+        Kss = torch.exp(-0.5 * torch.cdist(Z_source @ L_11, Z_source @ L_11))
+        Kst = torch.exp(-0.5 * torch.cdist(Z_source @ L_12, Z_target @ L_12))
+        Ktt = torch.exp(-0.5 * torch.cdist(Z_target @ L_22, Z_target @ L_22))
+
+        return torch.mean(Kss) + torch.mean(Ktt) - 2 * torch.mean(Kst)
+
+    def _F(self, W, X_source, y_source):
+        """Subspace learning objective function"""
+        torch = self.torch
+        base_method = self.base_method
+
+        if base_method == "pca":
+            cov = self._torch_cov(X_source)
+            loss = -torch.trace(W.T @ cov @ W)
+        elif base_method == "flda":
+            classes = torch.unique(y_source)
+            classes_means = torch.stack(
+                [torch.mean(X_source[y_source == c], dim=0) for c in classes]
+            )
+            classes_n_samples = torch.stack([torch.sum(y_source == c) for c in classes])
+            classes_means = classes_means * torch.sqrt(classes_n_samples).reshape(-1, 1)
+            S_W = self._torch_cov(classes_means)
+            S_B = self._torch_cov(X_source)
+            loss = torch.trace(W.T @ S_W @ W) / torch.trace(W.T @ S_B @ W)
+        elif base_method == "lpp":
+            # E is the Gaussian kernel if (y_source)_i == (y_source)_j and 0 otherwise
+            E = torch.exp(-0.5 * torch.cdist(X_source, X_source))
+            E = E * (y_source[:, None] == y_source[None, :])
+            D = torch.diag(torch.sum(E, dim=1))
+            loss = -torch.trace(W.T @ X_source.T @ (D - E) @ X_source @ W)
+
+        return loss
+
+    def fit(self, X, y=None, sample_domain=None, **kwargs):
+        """Fit adaptation parameters.
+
+        Parameters
+        ----------
+        X : array-like, shape (n_samples, n_features)
+            The source data.
+        y : array-like, shape (n_samples,)
+            The source labels.
+        sample_domain : array-like, shape (n_samples,)
+            The domain labels (same as sample_domain).
+
+        Returns
+        -------
+        self : object
+            Returns self.
+        """
+        try:
```

Comment:
I guess it is better to say so when instantiating the object (i.e. in the init) rather than when calling fit